### DISCUSSION TITLE
The future of Factory

### DISCUSSION THREAD

**Discussion by namit-chandwani** [aerogear/charmil#156] [Ideas]
As of now, this is how our Factory struct looks like:

```go
type Factory struct {
	// Type which defines the streams for the CLI
	IOStreams *iostreams.IOStreams
	// Logger provides functions for unified logging
	Logger logging.Logger
	// Localizer localizes the text in commands
	Localizer localize.Localizer
	// CfgHandler provides the fields required for managing config
	CfgHandler *config.CfgHandler
}
```

- In the ideal scenario, Factory should be initialized in the host CLI and passed down to its plugins. 
However, in the current setup of host and plugin examples, the Factory is being initialized in the plugins themselves.

- The only thing blocking us from achieving what we want is the Localizer.
Since each CLI's localizer configuration is unique, it doesn't make much sense to share it between host and their plugins.

- One way to fix this is by removing Localizer field from the Factory struct (as proposed by @wtrocki in https://github.com/aerogear/charmil/pull/155#issuecomment-877161611)

Let's come up with some more ideas for this and then decide which one to move forward with.

---

CC @aerogear/charmil-core

Ref: https://github.com/aerogear/charmil/pull/155#issuecomment-877165180, https://github.com/aerogear/charmil/pull/155#issuecomment-877161611, https://github.com/aerogear/charmil/pull/153

**Comment by wtrocki** [comment]
Localizer can be part of starter template

**Reply by namit-chandwani** [reply]
Yes, this seems to be the safest course of action at this stage.
Given that, if the need arises, we can always migrate it back to core.

**Reply by namit-chandwani** [reply]
@jackdelahunt @ankithans What are your thoughts on this?
If you're fine with this approach, I can make a PR for it before the config one.

**Reply by wtrocki** [reply]
Lets work on config :)
i18n is already in starter so there is nothing to work on right now.

**Reply by namit-chandwani** [reply]
Yeah, I'm working on config right now.

Last time, I had to apply  [this workaround](https://github.com/aerogear/charmil/pull/153/files#diff-c8139399572d6d2091e17f735d6ed81a6e0eb5fb53bcbe8d32712f2d795a14eeR36-R45) to avoid breaking any existing code.
So I was just trying to prevent it this time around (maybe by removing Localizer from the Factory struct).

But yes, this can be done later too

**Reply by wtrocki** [reply]
Remove i18n refactor what is needed and break stuff. No one is using core (yet)

**Reply by namit-chandwani** [reply]
Got it. Thanks!
